cowprotocol/tenderly-watch-tower
# feat: chain watcher

# Description
As we are removing Tenderly, we need to handle our own run loops for watching blocks / transaction logs.

# Changes

- [x] Add `warmUp` to initialise state from genesis, or last processed block.
- [x] Used rebuild logic from #29 for `warmUp`
- [x] Once `warmUp` is complete, flag as `inSync`, and commence block watching.
- [x] Implemented all in `ChainWatcher`, which runs per-chain, but we can handle many RPCs
- [x] All chains watched from the same instance.

## How to test

No unit tests available, however can observe functionality, such as with:

```bash
yarn ts-node ./src/watchtower.ts run --rpc http://erigon.dappnode:8545 --rpc http://nethermind-xdai.dappnode:8545 --deployment-block 17883049 --deployment-block 29389123 --page-size 0
```

**NOTE**: Substitute the above RPCs with the RPCs of your choice for the chains.

Observe synchronising of state. If Ctrl+C, observe `SIGINT` is caught, database is saved. Restart, and node resumes from last known synced block.

## Todo

- [x] Check for TODOs / dead code.

## Related Issues

Closes #46
